File: musesp_gameplay/menu_page.py
"""MenuPage —— 居中半透明菜单。"""

import pygame

from musesp_ui.components.core import Component, Constraintable, Direction
from musesp_ui.components.image_button import ImageButton
from musesp_ui.components.spacer import Spacer
from musesp_ui.router import Page


class MenuPage(Page):
    """不占满窗口的菜单页面，居中矩形区域，半透明白色背景。"""

    def full_shadow_promise(self) -> bool:
        return False

    def build(self) -> None:
        self._root.layout_direction = Direction.VERTICAL

        # 上 Spacer
        top = Spacer()
        top.v_constraint = Constraintable.MAXIMUM
        self._root.add_sub_component(top)

        # 按钮行
        row = Component()
        row.layout_direction = Direction.HORIZONTAL
        row.v_constraint = Constraintable.MINIMUM
        row.h_constraint = Constraintable.MINIMUM
        row.min_height = 36

        btns: list[tuple[str, str, object]] = [
            ("assets/ui/cancel_button.svg", "取消", self._on_cancel),
            ("assets/ui/continue_button.svg", "继续", lambda e: False),
            ("assets/ui/replay_button.svg", "重来", lambda e: False),
            ("assets/ui/exit_button.svg", "退出", self._on_exit),
        ]
        for i, (path, label, handler) in enumerate(btns):
            if i == 0:
                s = Spacer()
                s.h_constraint = Constraintable.MINIMUM
                s.min_width = 8
                row.add_sub_component(s)

            btn = ImageButton(path, text=label, width=130, height=36,
                              font_size=14)
            btn.h_constraint = Constraintable.MINIMUM
            btn.v_constraint = Constraintable.MINIMUM
            btn.min_width = 120
            btn.min_height = 36
            btn.bind_on_mouse_click(handler)
            if label == "取消":
                self._cancel_btn = btn
            row.add_sub_component(btn)

            s = Spacer()
            s.h_constraint = Constraintable.MINIMUM
            s.min_width = 8
            row.add_sub_component(s)

        self._root.add_sub_component(row)

        # 下 Spacer
        bot = Spacer()
        bot.v_constraint = Constraintable.MAXIMUM
        self._root.add_sub_component(bot)

    def prepare_layout(self) -> None:
        """根据窗口尺寸计算菜单区域。"""
        rw, rh = self._root.width, self._root.height
        mw = rw * 2 // 5
        mh = rh // 4
        self._menu_x = (rw - mw) // 2
        self._menu_y = (rh - mh) // 2
        self._menu_w = mw
        self._menu_h = mh
        self._root.x = self._menu_x
        self._root.y = self._menu_y
        self._root.width = self._menu_w
        self._root.height = self._menu_h
        self._root.layout()

        bg = pygame.Surface((self._menu_w, self._menu_h), pygame.SRCALPHA)
        bg.fill((255, 255, 255, 180))
        self._bg_surf = bg

    def draw(self, surface: pygame.Surface) -> None:
        self.draw_background(surface, (self._menu_x, self._menu_y,
                                       self._menu_w, self._menu_h),
                             self._bg_surf)
        super().draw(surface)

    def _on_cancel(self, event) -> bool:
        self._router.pop()
        return False

    def _on_exit(self, event) -> bool:
        from musesp.pages.music_list import MusicListPage
        self._router.clear_and_push(MusicListPage())
        return False

    def dispatch_event(self, event) -> bool:
        if event.type == pygame.KEYDOWN and event.key == pygame.K_ESCAPE:
            dummy = pygame.event.Event(pygame.MOUSEBUTTONUP, {"pos": (0, 0)})
            self._cancel_btn._emit("mouse_click", dummy)
            return False
        if event.type in (pygame.MOUSEMOTION, pygame.MOUSEBUTTONDOWN,
                          pygame.MOUSEBUTTONUP):
            mx, my = event.pos
            if not (self._menu_x <= mx < self._menu_x + self._menu_w
                    and self._menu_y <= my < self._menu_y + self._menu_h):
                return True
        return self._root.dispatch_event(event)

File: musesp_gameplay/src/menu_page.rs
use winit::event::{ElementState, WindowEvent};

use musesp_config::config::Config;
use musesp_ui::components::core::{ComponentBase, ComponentTrait, Constraintable, Direction};
use musesp_ui::components::image_button::ImageButton;
use musesp_ui::components::spacer::Spacer;
use musesp_ui::renderer::UIRenderer;
use musesp_ui::router::{AnyPage, NavAction, Page};

pub struct MenuPage {
    pub page: Page,
    menu_x: i32,
    menu_y: i32,
    menu_w: i32,
    menu_h: i32,
}


impl MenuPage {
    pub fn new() -> Self {
        MenuPage {
            page: Page::new(),
            menu_x: 0,
            menu_y: 0,
            menu_w: 0,
            menu_h: 0,
        }
    }
}

impl AnyPage for MenuPage {
    fn page(&self) -> &Page {
        &self.page
    }
    fn page_mut(&mut self) -> &mut Page {
        &mut self.page
    }
    fn full_shadow_promise(&self) -> bool {
        false
    }

    fn build(&mut self) {
        self.page.root.layout_direction = Direction::Vertical;

        let mut top = Spacer::new(0, 0);
        top.base.v_constraint = Constraintable::Maximum;
        self.page.root.children.push(top);

        let mut row = ComponentBase::new(0, 0, 0, 0);
        row.layout_direction = Direction::Horizontal;
        row.v_constraint = Constraintable::Minimum;
        row.h_constraint = Constraintable::Minimum;
        row.min_height = 36;

        let btns: [(&str, &str); 4] = [
            ("assets/ui/cancel_button.svg", "取消"),
            ("assets/ui/continue_button.svg", "继续"),
            ("assets/ui/replay_button.svg", "重来"),
            ("assets/ui/exit_button.svg", "退出"),
        ];

        let nav = self.page.nav.clone().unwrap();

        for (i, (path, label)) in btns.iter().enumerate() {
            if i == 0 {
                let mut s = Spacer::new(8, 0);
                s.base.h_constraint = Constraintable::Minimum;
                s.base.min_width = 8;
                row.children.push(s);
            }
            let mut btn = ImageButton::new(path, label, 0, 0, 130, 36, 14);
            btn.base.h_constraint = Constraintable::Minimum;
            btn.base.v_constraint = Constraintable::Minimum;
            btn.base.min_width = 120;
            btn.base.min_height = 36;

            let n = nav.clone();
            let lbl = *label;
            btn.base.bind_mouse_click(Box::new(move |_| {
                match lbl {
                    "取消" => {
                        let _ = n.send(NavAction::Pop);
                    }
                    "退出" => {
                        let _ = n.send(NavAction::Pop);
                        let _ = n.send(NavAction::Pop);
                    }
                    _ => {}
                }
                false
            }));
            row.children.push(btn);

            let mut s = Spacer::new(8, 0);
            s.base.h_constraint = Constraintable::Minimum;
            s.base.min_width = 8;
            row.children.push(s);
        }
        self.page
            .root
            .children
            .push(Box::new(row));

        let mut bot = Spacer::new(0, 0);
        bot.base.v_constraint = Constraintable::Maximum;
        self.page.root.children.push(bot);
    }

    fn prepare_layout(&mut self) {
        let rw = self.page.root.width;
        let rh = self.page.root.height;
        let mw = rw * 2 / 5;
        let mh = rh / 4;
        self.menu_x = (rw - mw) / 2;
        self.menu_y = (rh - mh) / 2;
        self.menu_w = mw;
        self.menu_h = mh;
        self.page.root.x = self.menu_x;
        self.page.root.y = self.menu_y;
        self.page.root.width = self.menu_w;
        self.page.root.height = self.menu_h;
        self.page.root.layout(None);
    }

    fn draw(&self, renderer: &mut UIRenderer) {
        renderer.draw_rect(
            self.menu_x,
            self.menu_y,
            self.menu_w,
            self.menu_h,
            (255, 255, 255, 180),
        );
        self.page.draw(renderer);
    }

    fn draw_debug(&self, renderer: &mut UIRenderer, config: &Config) {
        self.page.draw_debug(renderer, config);
    }

    fn dispatch_event(&mut self, event: &WindowEvent) {
        use winit::event::WindowEvent;
        use winit::keyboard::{KeyCode, PhysicalKey};

        if let WindowEvent::KeyboardInput {
            event: key_event, ..
        } = event
        {
            if key_event.state == ElementState::Pressed
                && key_event.physical_key == PhysicalKey::Code(KeyCode::Escape)
            {
                if let Some(ref nav) = self.page.nav {
                    let _ = nav.send(NavAction::Pop);
                }
                return;
            }
        }

        let in_menu = match event {
            WindowEvent::CursorMoved { position, .. } => {
                let px = position.x as i32;
                let py = position.y as i32;
                self.menu_x <= px
                    && px < self.menu_x + self.menu_w
                    && self.menu_y <= py
                    && py < self.menu_y + self.menu_h
            }
            WindowEvent::MouseInput { .. } => {
                let px = self.page.root.cursor_x as i32;
                let py = self.page.root.cursor_y as i32;
                self.menu_x <= px
                    && px < self.menu_x + self.menu_w
                    && self.menu_y <= py
                    && py < self.menu_y + self.menu_h
            }
            _ => true,
        };
        if !in_menu {
            return;
        }
        self.page.dispatch_event(event);
    }
}
